presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java

```diff
@@ -32,15 +41,36 @@
 {
     private final Node query;
 
+    private final boolean isValidationOnly;
+    private final Optional<Analysis> analysis;
+
+    private ImmutableList.Builder<List<TableColumn>> linkedBaseColumns;
+
     public MaterializedViewPlanValidator(Node query)
     {
+        this(query, true, Optional.empty());
+    }
+
+    public MaterializedViewPlanValidator(Node query, boolean isValidationOnly, Optional<Analysis> analysis)
+    {
+        checkArgument(isValidationOnly || analysis.isPresent(), "analysis must be set if");
+
         this.query = requireNonNull(query, "query is null");
+        this.isValidationOnly = isValidationOnly;
```

Comment:
Can you share more light on why we need this?

Reply:
actually ignore this, I got the intention, I will look more in depth the logic.

Reply:
Right, it's a flag whether the class is used to validate the MV shape only or is used to compute the LinkedBaseColumns. To compute the LinkedBaseColumns, it requires passing the analysis object and thus can only be called after we get the entire analysis object.

I see we right now need to call this class, MaterializedviewPlanValidator, in StatementAnalyzer.visitTable() -> processMaterializedView() -> validateMaterializedViewQueryPlan(), https://github.com/prestodb/presto/blob/692ca3de145c5664237315bc18e45b6fd7e7d7d0/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L702. That's where we haven't finish visiting and don't have the entire analysis object yet. There we only need MV shape validation and don't need the analysis object. To address that case, I'm adding this isValidatationOnly flag and set it to True in the case.